README.md
# python-ldap-test

Tool for testing code speaking with LDAP server. Allows to easily configure and run 
an embedded, in-memory LDAP server. Uses UnboundID LDAP SDK through Py4J.
Requires Java runtime on the system path to run the server.

## Installation

With `pip`:

    pip install python-ldap-test


When installing from source:

    git clone https://github.com/zoldar/python-ldap-test
    cd python-ldap-test
    python setup.py install # you may need root privileges if installing system-wide

## Usage

Example library usage with Python ldap client.

    import ldap3

    from ldap_test import LdapServer

    server = LdapServer()

    try:
        server.start()

        dn = server.config['bind_dn']
        pw = server.config['password']

        srv = ldap3.Server('localhost', port=server.config['port'])
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)

        base_dn = server.config['base']['dn']
        search_filter = '(objectclass=domain)'
        attrs = ['dc']

        conn.search(base_dn, search_filter, attributes=attrs)

        print conn.response
        # [{
        #    'dn': 'dc=example,dc=com',
        #    'raw_attributes': {'dc': [b'example']},
        #    'attributes': {'dc': ['example']},
        #    'type': 'searchResEntry'
        # }]
    finally:
        server.stop()

Another example with non-standard settings:

    import ldap3

    from ldap_test import LdapServer

    server = LdapServer({
        'port': 3333,
        'bind_dn': 'cn=admin,dc=zoldar,dc=net',
        'password': 'pass1',
        'base': {'objectclass': ['domain'],
                 'dn': 'dc=zoldar,dc=net',
                 'attributes': {'dc': 'zoldar'}},
        'entries': [
            {'objectclass': 'domain',
             'dn': 'dc=users,dc=zoldar,dc=net',
             'attributes': {'dc': 'users'}},
            {'objectclass': 'organization',
             'dn': 'o=foocompany,dc=users,dc=zoldar,dc=net',
             'attributes': {'o': 'foocompany'}},
        ]
    })

    try:
        server.start()

        dn = "cn=admin,dc=zoldar,dc=net"
        pw = "pass1"

        srv = ldap3.Server('localhost', port=3333)
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)

        base_dn = 'dc=zoldar,dc=net'
        search_filter = '(objectclass=organization)'
        attrs = ['o']

        conn.search(base_dn, search_filter, attributes=attrs)

        print conn.response
        # [{
        #    'dn': 'o=foocompany,dc=users,dc=zoldar,dc=net',
        #    'raw_attributes': {'o': [b'foocompany']},
        #    'attributes': {'o': ['foocompany']},
        #    'type': 'searchResEntry'
        # }]
    finally:
        server.stop()

And, finally, an example of running multiple LDAP servers:

    import ldap3

    from ldap_test import LdapServer

    servers = {}

    try:
        for sid in (1, 2):
            domain = 'example{0}'.format(sid)
            servers[sid] = LdapServer({
                'port': 10389 + (sid * 1000),
                'bind_dn': 'cn=admin,dc={0},dc=com'.format(domain),
                'base': {
                    'objectclass': ['domain'],
                    'dn': 'dc={0},dc=com'.format(domain),
                    'attributes': {'dc': domain}
                },
            })
            servers[sid].start()

        search_filter = '(objectclass=domain)'
        attrs = ['dc']

        # server1
        dn = servers[1].config['bind_dn']
        pw = servers[1].config['password']
        base_dn = servers[1].config['base']['dn']
        port = servers[1].config['port']

        srv = ldap3.Server('localhost', port=port)
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)
        conn.search(base_dn, search_filter, attributes=attrs)

        print conn.response
        # [{
        #    'dn': 'dc=example1,dc=com',
        #    'raw_attributes': {'dc': [b'example1']},
        #    'attributes': {'dc': ['example1']},
        #    'type': 'searchResEntry'
        # }]

        conn.unbind()

        # server2
        dn = servers[2].config['bind_dn']
        pw = servers[2].config['password']
        base_dn = servers[2].config['base']['dn']
        port = servers[2].config['port']

        srv = ldap3.Server('localhost', port=port)
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)
        conn.search(base_dn, search_filter, attributes=attrs)

        print conn.response
        # [{
        #    'dn': 'dc=example2,dc=com',
        #    'raw_attributes': {'dc': [b'example2']},
        #    'attributes': {'dc': ['example2']},
        #    'type': 'searchResEntry'
        # }]

        conn.unbind()
    finally:
        for server in servers.values():
            server.stop()

The initial server configuration is represented by a simple dict, which may
contain one or more optional parameters:

- `port` - a port on which the LDAP server will listen
- `bind_dn` - bind DN entry for authentication
- `password` - authentication password
- `base` - base DN entry
- `entries` - a list of dicts representing intially loaded entries 
   in the database. `attributes` are optional here
- `ldifs` - a list of strings representing file paths to the LDIF files to load
  on start, like `..., 'ldifs': ['path/to/file1.ldif', 'path/to/file2.ldif'], ...`

The format of entry in `entries` as well as `base` is following:

    {'dn': 'o=some,dc=example,dc=com', # DN identifying the entry
     'objectclass': ['top', 'organization'], # objectclass may be either a 
                                             # string in case of a single 
                                             # class or a list of classes
     'attributes': {        # attributes are optional
        'o': 'some'         # every attribute may have either a single value
                            # or multiple values in a list like
                            # 'ou': ['Value1', 'Value2', ...]
     }
    }
    
## MacOS
For some reason, while running on MacOS, you can experience problems if the JVM doesn't start quickly enough for the py4j gateway to connect, and it goes into an infinite hang.

If you're experiencing this problem, you can set an interval between the JVM startup and the py4j gateway by passing 'java_delay=n' to LdapServer() where 'n' is the number of seconds to wait.  Typically a wait of even 1 second is enough for java to spin up and be ready for the gateway.

To be clear, the following:

    server = LdapServer(config={...}, java_delay=1)
    
will cause a 1 second delay between starting the JVM and creating the py4j gateway, and all should be well.  

In tests, a delay of even 0.5 seconds can be enough, though 0.1 seconds is not.  The exact cause of this problem is unknown.  More information on this 'feature' while running on a Mac is welcome.

## Running Java gateway and proxy on non-standard ports

When there's a necessity to run proxy and gateway on ports different from the default ones
(25333 for Java gateway and 25334 for proxy), the `LdapServer` may be instantiated with
custom ones, passed explicitly to the constructor:

```
server = LdapServer({...}, java_gateway_port=26333, python_proxy_port=26334)
```

This can be useful when several test runs are done in parallel on a single system.

## Reporting issues

Any issues (be it bugs, feature requests or anything else) can be reported through project's [GitHub issues page](https://github.com/zoldar/python-ldap-test/issues).

## Contributors

- John Kristensen ([https://github.com/jerrykan](https://github.com/jerrykan))
- Kevin Rasmussen ([https://github.com/krasmussen](https://github.com/krasmussen))
- Pedro Algarvio ([https://github.com/s0undt3ch](https://github.com/s0undt3ch))
- Nik Ogura ([https://github.com/nikogura](https://github.com/nikogura))

## License

Copyright © 2016 Adrian Gruntkowski

Distributed under the MIT License.


CHANGES.txt
v0.3.1, 2017-05-19 -- Introduced optional delay between JVM and py4j gateway start to solve issue # 14
v0.3.0, 2016-09-11 -- Support for configuring custom ports for gateway and proxy
v0.2.2, 2016-07-27 -- Update py4j to make library work with recent python versions
v0.2.1, 2015-12-30 -- More informative error on missing java executable
v0.2.0, 2014-09-23 -- Support for running multiple LDAP servers
v0.1.1, 2014-09-19 -- Compatibility with Python3
v0.0.8, 2013-09-20 -- Fixed grave issue while loading ldifs. Disabled schema validation.
v0.0.7, 2013-07-05 -- First release with reasonable documentation.


ldap_test/test/tests.py
import sys
import unittest
from os import path

sys.path.insert(0, path.abspath(path.join(path.dirname(__file__), '..', '..')))

import ldap3

from ldap_test import LdapServer


class LdapServerTest(unittest.TestCase):
    def test_default(self):
        server = LdapServer()
        server.start()

        dn = server.config['bind_dn']
        pw = server.config['password']

        srv = ldap3.Server('localhost', port=server.config['port'])
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)

        base_dn = server.config['base']['dn']
        search_filter = '(objectclass=domain)'
        attrs = ['dc']

        conn.search(base_dn, search_filter, attributes=attrs)

        self.assertEqual(conn.response, [{
            'dn': 'dc=example,dc=com',
            'raw_attributes': {'dc': [b'example']},
            'attributes': {'dc': ['example']},
            'type': 'searchResEntry'
        }])

        conn.unbind()
        server.stop()

    def test_default_with_delay(self):
        server = LdapServer(java_delay=1)
        server.start()

        dn = server.config['bind_dn']
        pw = server.config['password']

        srv = ldap3.Server('localhost', port=server.config['port'])
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)

        base_dn = server.config['base']['dn']
        search_filter = '(objectclass=domain)'
        attrs = ['dc']

        conn.search(base_dn, search_filter, attributes=attrs)

        self.assertEqual(conn.response, [{
            'dn': 'dc=example,dc=com',
            'raw_attributes': {'dc': [b'example']},
            'attributes': {'dc': ['example']},
            'type': 'searchResEntry'
        }])

        conn.unbind()
        server.stop()

    def test_config(self):
        server = LdapServer({
            'port': 3333,
            'bind_dn': 'cn=admin,dc=zoldar,dc=net',
            'password': 'pass1',
            'base': {
                'objectclass': ['domain'],
                'dn': 'dc=zoldar,dc=net',
                'attributes': {'dc': 'zoldar'}
            },
            'entries': [{
                'objectclass': ['domain'],
                'dn': 'dc=users,dc=zoldar,dc=net',
                'attributes': {'dc': 'users'}
            }, {
                'objectclass': ['organization'],
                'dn': 'o=foocompany,dc=users,dc=zoldar,dc=net',
                'attributes': {'o': 'foocompany'}
            }],
        })
        server.start()

        dn = "cn=admin,dc=zoldar,dc=net"
        pw = "pass1"

        srv = ldap3.Server('localhost', port=3333)
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)

        base_dn = 'dc=zoldar,dc=net'
        search_filter = '(objectclass=organization)'
        attrs = ['o']

        conn.search(base_dn, search_filter, attributes=attrs)

        self.assertEqual(conn.response, [{
            'dn': 'o=foocompany,dc=users,dc=zoldar,dc=net',
            'raw_attributes': {'o': [b'foocompany']},
            'attributes': {'o': ['foocompany']},
            'type': 'searchResEntry'
        }])

        conn.unbind()
        server.stop()

    def test_non_standard_objects(self):
        server = LdapServer({
            'port': 3333,
            'bind_dn': 'cn=admin,dc=zoldar,dc=net',
            'password': 'pass1',
            'base': {
                'objectclass': ['domain'],
                'dn': 'dc=zoldar,dc=net',
                'attributes': {'dc': 'zoldar'}
            },
            'entries': [{
                'objectclass': ['domain'],
                'dn': 'dc=users,dc=zoldar,dc=net',
                'attributes': {'dc': 'users'}
            }, {
                'objectclass': ['foo'],
                'dn': 'bar=foovar,dc=users,dc=zoldar,dc=net',
                'attributes': {'bar': 'foovar'}
            }],
        })
        server.start()

        dn = "cn=admin,dc=zoldar,dc=net"
        pw = "pass1"

        srv = ldap3.Server('localhost', port=3333)
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)

        base_dn = 'dc=zoldar,dc=net'
        search_filter = '(objectclass=foo)'
        attrs = ['bar']

        conn.search(base_dn, search_filter, attributes=attrs)

        self.assertEqual(conn.response, [{
            'dn': 'bar=foovar,dc=users,dc=zoldar,dc=net',
            'raw_attributes': {'bar': [b'foovar']},
            'attributes': {'bar': ['foovar']},
            'type': 'searchResEntry'
        }])

        conn.unbind()
        server.stop()

    def test_multiple_instances(self):
        servers = {}
        for sid in (1, 2):
            domain = 'example{0}'.format(sid)
            servers[sid] = LdapServer({
                'port': 10389 + (sid * 1000),
                'bind_dn': 'cn=admin,dc={0},dc=com'.format(domain),
                'base': {
                    'objectclass': ['domain'],
                    'dn': 'dc={0},dc=com'.format(domain),
                    'attributes': {'dc': domain}
                },
            })
            servers[sid].start()

        search_filter = '(objectclass=domain)'
        attrs = ['dc']

        # server1
        dn = servers[1].config['bind_dn']
        pw = servers[1].config['password']
        base_dn = servers[1].config['base']['dn']
        port = servers[1].config['port']

        srv = ldap3.Server('localhost', port=port)
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)
        conn.search(base_dn, search_filter, attributes=attrs)

        self.assertEqual(conn.response, [{
            'dn': 'dc=example1,dc=com',
            'raw_attributes': {'dc': [b'example1']},
            'attributes': {'dc': ['example1']},
            'type': 'searchResEntry'
        }])

        conn.unbind()

        # server2
        dn = servers[2].config['bind_dn']
        pw = servers[2].config['password']
        base_dn = servers[2].config['base']['dn']
        port = servers[2].config['port']

        srv = ldap3.Server('localhost', port=port)
        conn = ldap3.Connection(srv, user=dn, password=pw, auto_bind=True)
        conn.search(base_dn, search_filter, attributes=attrs)

        self.assertEqual(conn.response, [{
            'dn': 'dc=example2,dc=com',
            'raw_attributes': {'dc': [b'example2']},
            'attributes': {'dc': ['example2']},
            'type': 'searchResEntry'
        }])

        conn.unbind()

        for server in servers.values():
            server.stop()


if __name__ == '__main__':
    unittest.main()


ldap_test/server.py
#!/usr/bin/env python

import logging
import os
import time
import subprocess
import atexit
from distutils.spawn import find_executable

from py4j.java_gateway import JavaGateway, GatewayClient, GatewayConnection, GatewayParameters
from py4j.protocol import Py4JNetworkError

try:
    basestring
except NameError:
    basestring = (str, bytes)


CONFIG_PARAMS = ('port', 'base', 'entries', 'ldifs', 'bind_dn', 'password')
DEFAULT_CONFIG = {
    'port': 10389,
    'bind_dn': 'cn=admin,dc=example,dc=com',
    'password': 'password',
    'base': {'objectclass': ['domain'],
             'dn': 'dc=example,dc=com',
             'attributes': {'dc': 'example'}}
}

DEFAULT_GATEWAY_PORT = 25333
DEFAULT_PYTHON_PROXY_PORT = 25334
PACKAGE_DIR = os.path.dirname(os.path.abspath(__file__))
JVM_SERVER_BIN = os.path.join(
    PACKAGE_DIR,
    "ldap-test-server-0.0.4-SNAPSHOT-jar-with-dependencies.jar")

logging.basicConfig(level=logging.INFO)
log = logging.getLogger('ldap_test')

SERVER_PROCESS = None
JVM_GATEWAY = None


def server_cleanup():
    global SERVER_PROCESS, JVM_GATEWAY
    if JVM_GATEWAY:
        JVM_GATEWAY.shutdown()
    if SERVER_PROCESS:
        SERVER_PROCESS.kill()


atexit.register(server_cleanup)


def run_jvm_gateway(gateway_port=DEFAULT_GATEWAY_PORT,
                    python_proxy_port=DEFAULT_PYTHON_PROXY_PORT):
    try:
        log.info('Starting the JavaGateway on python_proxy_port %s', python_proxy_port)
        return JavaGateway(gateway_client=SlowGatewayClient(port=gateway_port),
                           python_proxy_port=python_proxy_port)
    except Py4JNetworkError:
        log.error("Failed to connect!")
        raise


def run_jvm_server(gateway_port=DEFAULT_GATEWAY_PORT):
    if not os.path.isfile(JVM_SERVER_BIN):
        raise Exception("%s is missing!" % (JVM_SERVER_BIN,))

    jre_executable = find_executable("java")

    if jre_executable is None:
        raise Exception("'java' executable not found in system path!")

    try:
        return subprocess.Popen("exec %s -jar %s --port %s" % (
            jre_executable,
            JVM_SERVER_BIN,
            gateway_port), shell=True)
    except OSError as e:
        log.error("Failed to run JVM server because: %s" % (e,))
        raise


class SlowGatewayClient(GatewayClient):
    def _create_connection(self):
        while True:
            parameters = GatewayParameters(address=self.address,
                                           port=self.port,
                                           auto_close=self.auto_close,
                                           eager_load=True)
            connection = MuffledGatewayConnection(parameters, self.gateway_property)
            connection_success = False
            try:
                connection.start()
                connection_success = True
            except Py4JNetworkError:
                pass
            except (KeyboardInterrupt, SystemExit):
                break
            if connection_success:
                break
            time.sleep(0.1)
        return connection


class MuffledGatewayConnection(GatewayConnection):
    def start(self):
        try:
            log.debug('%s is trying to connect to %s:%d', self.__class__.__name__, self.address, self.port)
            self.socket.connect((self.address, self.port))
            self.is_connected = True
            self.stream = self.socket.makefile('rb', 0)
            log.debug('%s is now connected to the Java server', self.__class__.__name__)
        except Exception:
            msg = 'An error occurred while trying to connect to the Java '\
                            'server'
            raise Py4JNetworkError(msg)


class ConfigBuilder(object):
    def __init__(self, gateway):
        self.gateway = gateway
        self.builder = self.gateway.jvm.net.zoldar.ldap.testserver.LdapBuilder

    def build_from(self, config_map=None):
        config_map = config_map or {}
        config_map = dict(DEFAULT_CONFIG, **config_map)
        config = self.builder.config()

        for param in CONFIG_PARAMS:
            if param in config_map:
                getattr(self, "_set_%s" % (param,))(config, config_map[param])

        return config_map, config.build()

    def _set_port(self, config, port):
        config.port(port)

    def _set_base(self, config, entry_map):
        config.base(self._map_to_entry(entry_map))

    def _set_entries(self, config, entries):
        config.entries(self._list_to_array(
            self.gateway.jvm.com.github
                .trevershick.test.ldap.annotations.LdapEntry,
            [self._map_to_entry(x) for x in entries]
        ))

    def _set_ldifs(self, config, ldifs):
        config.ldifs(self._list_to_array(
            self.gateway.jvm.com.github.trevershick.test.ldap.annotations.Ldif,
            [self.builder.ldif(x) for x in ldifs]
        ))

    def _set_bind_dn(self, config, bind_dn):
        config.bindDn(bind_dn)

    def _set_password(self, config, password):
        config.password(password)

    def _map_to_entry(self, entry):
        dn = entry['dn']
        objectclass = entry['objectclass']
        attributes = entry.get('attributes') or {}

        if isinstance(objectclass, basestring):
            objectclass = [objectclass]

        attr_objects = []
        for attr_name in attributes:
            value = attributes[attr_name]
            if isinstance(value, basestring):
                value = [value]
            attr_objects.append(self.builder.attribute(
                attr_name,
                self._list_to_array(self.gateway.jvm.String, value)
            ))

        return self.builder.entry(
            dn,
            self._list_to_array(self.gateway.jvm.String, objectclass),
            self._list_to_array(
                self.gateway.jvm.com.github
                    .trevershick.test.ldap.annotations.LdapAttribute,
                attr_objects)
        )

    def _list_to_array(self, value_type, lst):
        array = self.gateway.new_array(value_type, len(lst))
        for idx, value in enumerate(lst):
            array[idx] = value
        return array


class LdapServer(object):
    def __init__(self, config=None,
                 java_gateway_port=DEFAULT_GATEWAY_PORT,
                 python_proxy_port=DEFAULT_PYTHON_PROXY_PORT, java_delay=None):
        global SERVER_PROCESS, JVM_GATEWAY

        if SERVER_PROCESS is None:
            SERVER_PROCESS = run_jvm_server(java_gateway_port)

        # Added to introduce a delay between starting the SERVER_PROCESS and the JVM_GATEWAY if desired.
        # This seems to be a problem on some MacOS systems, and without it you end up with an infinite hang.
        if java_delay:
            time.sleep(java_delay)

        if JVM_GATEWAY is None:
            JVM_GATEWAY = run_jvm_gateway(java_gateway_port, python_proxy_port)

        self.server = JVM_GATEWAY.entry_point
        self.config, self._config_obj = \
            ConfigBuilder(JVM_GATEWAY).build_from(config)
        self.server_id = self.server.create(self._config_obj)

    def start(self):
        self.server.start(self.server_id)

    def stop(self):
        self.server.stop(self.server_id)
